app/services/activity/updater.rb

```diff
@@ -0,0 +1,180 @@
+class Activity
+  class Updater
+    include DateHelper
+    include CodelistHelper
+    include ActivityHelper
+
+    attr_reader :activity, :params
+
+    def initialize(activity:, params:)
+      @activity = activity
+      @params = params
+    end
+
+    def update(step)
+      updater = "set_#{step}"
+      if respond_to?(updater, :include_private)
+        __send__(updater)
+      else
+        set_for_step(step)
+      end
+    end
+
+    private
+
+    def set_for_step(step)
```

Comment:
To be consistent, should this be called `set_step` ? Although I'm confused what this is used for.

Also, would it be useful to have a test that checks that `ActivityUpdater` has a `set_` method for every step? I note that `set_programme_status` is missing, for example.

I'm wondering if the `else` is a mistake here, as it's swallowing `respond_to?` failures.

Reply:
I've also wondered why, since all users who can manage an activity can access all steps, we can't globally do...

`activity.assign_attributes(activity_params)`

...rather than carefully making sure the form only contains the fields from the step you're currently on, which should always be the case, unless the user hacks the form, which would be pointless.

Reply:
`set_for_step` is a generic catch-all method for any steps that don't require any special actions, so we can just set the relevant params and move on. That said, we could probably just set the whole params and move on, as you say. I'll have a crack at that and see if anything else falls out

Reply:
Oh, also, just running `activity.assign_attributes(activity_params)` gives an `ActiveModel::ForbiddenAttributesError` in some cases, presumably because there's some stuff submitted in forms that aren't present in the model.

Reply:
Maybe keep the individual `assign_attributes` for now so it's a straight refactor.  

I've realised I completely misread `step => params_for(step)` as the equivalent of `"step" => params_for("step")` instead of `step` being a variable for some reason or another 🤯  (!)

I still think having a distinct method name to keep it separate from the dynamic `step_*` methods might be useful: `fallback_set_for_step`

Reply:
Cool, I've gone for `assign_attributes_for_step` and put it seperate from the `step_*` methods to make it clearer